pkg/register/component.go

```diff
@@ -0,0 +1,202 @@
+	UpdateUser(accessToken string, realmName, userID string, user kc.UserRepresentation) error
+	GetUsers(accessToken string, reqRealmName, targetRealmName string, paramKV ...string) (kc.UsersPageRepresentation, error)
+	ExecuteActionsEmail(accessToken string, realmName string, userID string, actions []string, paramKV ...string) error
+}
+
+// ConfigurationDBModule is the interface of the configuration module.
+type ConfigurationDBModule interface {
+	GetConfiguration(context.Context, string) (dto.RealmConfiguration, error)
+}
+
+// Component is the register component interface.
+type Component interface {
+	RegisterUser(ctx context.Context, realmName string, user apiregister.User) (string, error)
+}
+
+// Component is the management component.
+type component struct {
+	realm          string
+	keycloakClient KeycloakClient
+	tokenProvider  keycloak.OidcTokenProvider
+	usersDBModule  UsersDBModule
+	configDBModule ConfigurationDBModule
+	eventsDBModule database.EventsDBModule
+	logger         internal.Logger
+}
+
+// NewComponent returns the management component.
+func NewComponent(realm string, keycloakClient KeycloakClient, tokenProvider keycloak.OidcTokenProvider, usersDBModule UsersDBModule, configDBModule ConfigurationDBModule, eventsDBModule database.EventsDBModule, logger internal.Logger) Component {
+	return &component{
+		realm:          realm,
+		keycloakClient: keycloakClient,
+		tokenProvider:  tokenProvider,
+		usersDBModule:  usersDBModule,
+		configDBModule: configDBModule,
+		eventsDBModule: eventsDBModule,
+		logger:         logger,
+	}
+}
+
+func (c *component) reportEvent(ctx context.Context, apiCall string, values ...string) {
+	errEvent := c.eventsDBModule.ReportEvent(ctx, apiCall, "back-office", values...)
+	if errEvent != nil {
+		//store in the logs also the event that failed to be stored in the DB
+		internal.LogUnrecordedEvent(ctx, c.logger, apiCall, errEvent.Error(), values...)
+	}
+}
+
+func (c *component) RegisterUser(ctx context.Context, realmName string, user apiregister.User) (string, error) {
+	// Validate input request
+	var err = user.Validate()
+	if err != nil {
+		c.logger.Info(ctx, "err", err.Error())
+		return "", err
+	}
+
+	// Get Realm configuration from database
+	var realmConf dto.RealmConfiguration
+	realmConf, err = c.configDBModule.GetConfiguration(ctx, realmName)
+	if err != nil {
+		c.logger.Info(ctx, "msg", "Can't get realm configuration from database", "err", err.Error())
+		return "", err
+	}
+
+	// Get an OIDC token to be able to request Keycloak
+	var accessToken string
+	accessToken, err = c.tokenProvider.ProvideToken(ctx)
+	if err != nil {
+		c.logger.Warn(ctx, "msg", "Can't get OIDC token", "err", err.Error())
+		return "", err
+	}
+
+	// Registering should be disallowed if an enabled user already exists with the same email
+	var kcUser *kc.UserRepresentation
+	kcUser, err = c.checkExistingUser(ctx, accessToken, user)
```

Comment:
In a very unlikely corner case, it may happen the user is not existing at this time but created just after. It will lead to a loop for username generation (hopefully only 10 times ;) ).